sympy/functions/combinatorial/factorials.py

```diff
@@ -1057,3 +1058,93 @@ def _eval_is_nonnegative(self):
                 return True
             elif k.is_even is False:
                 return  False
+
+class multinomial(CombinatorialFunction):
+    r"""Implementation of the multinomial coefficient. They can be expressed as
+    a product of binomial coefficients or of factorials:
+
+    .. math:: \frac{(k_1+k_2+\cdots+k_m)!}{k_1!\, k_2! \cdots k_m!} = {k_1\choose k_1}{k_1+k_2\choose k_2}\cdots{k_1+k_2+\cdots+k_m\choose k_m}
+
+    The multinomial coefficients have a direct combinatorial interpretation,
+    as the number of ways of depositing n distinct objects into m distinct bins,
+    with k1 objects in the first bin, k2 objects in the second bin, and so on.
+
+    To expand the multinomial when `k` contains a symbol, use either
+    ``expand_func()`` or ``expand(func=True)``. The former will keep
+    the polynomial in factored form while the latter will expand the
+    polynomial itself. See examples for details.
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, Rational, multinomial, expand_func, symbols
+    >>> n = Symbol('n', integer=True, positive=True)
+
+    >>> multinomial(15, 8)
+    490314
+
+    >>> multinomial(n, -1)
+    0
+
+    Comparing Coefficents generated with the multinomial function:
+
+    >>> a, b, c = symbols('a b c')
+    >>> ((a+b+c)**3).expand(func=True)
+    a**3 + 3*a**2*b + 3*a**2*c + 3*a*b**2 + 6*a*b*c + 3*a*c**2 + b**3 + 3*b**2*c + 3*b*c**2 + c**3
+
+    The Coefficient when of 2 1 0 corresponding to the second term is 3
+
+    >>> multinomial(2,1,0)
+    3
+
+    >>> multinomial(Rational(5, 4), 3)
+    gamma(21/4)/(6*gamma(9/4))
+
+    >>> multinomial(n, 3)
+    gamma(n + 4)/(6*gamma(n + 1))
+
+    >>> multinomial(n, 3).expand(func=True)
+    n**3/6 + n**2 + 11*n/6 + 1
+
+    >>> expand_func(multinomial(n, 3))
+    (n + 1)*(n + 2)*(n + 3)/6
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Multinomial_theorem#Multinomial_coefficients
+
+    .. [2] http://functions.wolfram.com/GammaBetaErf/Multinomial/02/0001/
+
+    """
+
+    @classmethod
+    def eval(cls, *k):
+        k = [sympify(x) for x in k]
+        numerator = Add(*k)
+        if all(x.is_Integer for x in k) and numerator > 0:
+            return factorial(numerator)/Mul(*(factorial(x) for x in k))
```

Comment:
Is factorial cached? If not then we could use `Mul(*[facorial(k)**v for ki,v in multiset(k).items()])`

Reply:
could you gimme some pointers on this ? ... cause I don't know.

Reply:
I looked [here](https://github.com/sympy/sympy/blob/516fa83e69caf1e68306cfc912a13f36c434d51c/sympy/functions/combinatorial/factorials.py#L37) and it looks like some values used to do the calculation are cached and the primes used to compute the values is cached so perhaps this is not needed. My guess is that for the general use case it will slow things down with another function call.

In addition, since the answer must be an integer when the input is an integer, it is not necessary to find the product of the individual factorials:
```python
rv = factorial(numerator)
for ki in k:
    rv //= factorial(ki)
```
But [this](https://stackoverflow.com/questions/22301115/efficient-matlab-implementation-of-multinomial-coefficient) suggests that we can do better if we just " write your multinomial coefficient as a product of binomial coefficients and use a built-in method to compute binomial coefficients". (And we have `nC` [here](https://github.com/sympy/sympy/blob/516fa83e69caf1e68306cfc912a13f36c434d51c/sympy/functions/combinatorial/numbers.py#L1624).)